src/pybind/mgr/dashboard/controllers/host.py

```diff
@@ -183,6 +203,8 @@ def get_hosts(sources=None):
         orch = OrchClient.instance()
         if orch.available():
             return merge_hosts_by_hostname(ceph_hosts, orch.hosts.list())
+    for host in ceph_hosts:
+        host['service_instances'] = populate_service_instances(host['hostname'], host['services'])
     return ceph_hosts
```

Comment:
I find this a bit intricate, since `merge_hosts_by_hostname()` already invokes this `populate_service_instances()` so why not making everything a bit more explicit and in a single codeflow (while in Python is acceptable to have multiple `return` statements, that usually leads to functions containing multiple mutually exclusive code paths, which complicates understanding the program logic):

```py
HOST_EXTRA_ORCH_FIELDS = {
        'addr': '',
        'labels': [],
        'sources': {
            'ceph': True,
            'orchestrator': False
        },
        'status': ''
    }

# Let's have clear, separate arguments rather than packed ones...
def get_hosts(from_ceph=True, from_orchestrator=True):
    """
    Get hosts from various sources.
    """
    ceph_hosts = []
    if from_ceph:
      ceph_hosts = [merge_dicts(server, HOST_EXTRA_ORCH_FIELDS)
          for server in mgr.list_servers()]

    if from_orchestrator:
        orch = OrchClient.instance()
        if orch.available():
            ceph_hosts = merge_hosts_by_hostname(ceph_hosts, orch.hosts.list())

    # Let's do this here just once (instead twice in 2 different paths)
    for host in ceph_hosts:
        host['service_instances'] = populate_service_instances(host['hostname'], host['services'])

    return ceph_hosts
```

Reply:
I tried this but it gave me test errors in the `test_host.py` file with some error. And fixing them gave me another error, which made me think that its going out of the scope for this PR. So if its okay I would go as it is since this is already an improvement over the exisiting behaviour (performance wise).

Reply:
ok, np